### PR TITLE
opae: add dlf driver support to Errors and Metrics

### PR DESCRIPTION
**Pull request by anandaravuri**
- Temperature
  - Power
  - Perf
  - FMR and Port Errors

### CODE REVIEW

File: libopae/plugins/xfpga/metrics/metrics_utils.c [hunk 1]
```diff
@@ -233,41 +231,37 @@ fpga_result enum_thermalmgmt_metrics(fpga_metric_vector *vector,
...
 		}
 
-		result = add_metric_vector(vector, *metric_num, THERLGMT, THERLGMT, sysfs_path, dirent->d_name, metric_sysfs, metric_data.metric_units,
-								FPGA_METRIC_DATATYPE_INT, FPGA_METRIC_TYPE_THERMAL, hw_type, 0);
+		result = add_metric_vector(vector, *metric_num, THERLGMT, THERLGMT, sysfspath, (dir_name + 1), pglob.gl_pathv[i], metric_data.metric_units,
+			FPGA_METRIC_DATATYPE_INT, FPGA_METRIC_TYPE_THERMAL, hw_type, 0);
 		if (result != FPGA_OK) {
-			closedir(dir);
 			FPGA_MSG("Failed to add metrics");
 			return result;
```
Comment (tswhison) [inline libopae/plugins/xfpga/metrics/metrics_utils.c]: pglob is leaked here.

File: libopae/plugins/xfpga/metrics/metrics_utils.c [hunk 2]
```diff
@@ -233,41 +231,37 @@ fpga_result enum_thermalmgmt_metrics(fpga_metric_vector *vector,
...
 		}
-
 		*metric_num = *metric_num + 1;
-
 	}
-	closedir(dir);
+
 	return result;
```
Comment (tswhison) [inline libopae/plugins/xfpga/metrics/metrics_utils.c]: pglob is leaked here.

File: libopae/plugins/xfpga/metrics/metrics_utils.c [hunk 3]
```diff
@@ -293,40 +285,38 @@ fpga_result enum_powermgmt_metrics(fpga_metric_vector *vector,
...
 		}
 
-		result = add_metric_vector(vector, *metric_num, PWRMGMT, PWRMGMT, sysfs_path, dirent->d_name, metric_sysfs, metric_data.metric_units,
-									FPGA_METRIC_DATATYPE_INT, FPGA_METRIC_TYPE_POWER, hw_type, 0);
+		result = add_metric_vector(vector, *metric_num, PWRMGMT, PWRMGMT, sysfspath, (dir_name + 1), pglob.gl_pathv[i], metric_data.metric_units,
+			FPGA_METRIC_DATATYPE_INT, FPGA_METRIC_TYPE_POWER, hw_type, 0);
 		if (result != FPGA_OK) {
 			FPGA_MSG("Failed to add metrics");
-			closedir(dir);
 			return result;
```
Comment (tswhison) [inline libopae/plugins/xfpga/metrics/metrics_utils.c]: pglob is leaked here.

File: libopae/plugins/xfpga/metrics/metrics_utils.c [hunk 4]
```diff
@@ -293,40 +285,38 @@ fpga_result enum_powermgmt_metrics(fpga_metric_vector *vector,
...
 		}
-
 		*metric_num = *metric_num + 1;
 	}
-	closedir(dir);
+
+
 	return result;
```
Comment (tswhison) [inline libopae/plugins/xfpga/metrics/metrics_utils.c]: pglob is leaked here.

File: libopae/plugins/xfpga/metrics/metrics_utils.c [hunk 5]
```diff
@@ -826,19 +817,32 @@ fpga_result enum_fpga_metrics(fpga_handle handle)
...
-			result = enum_powermgmt_metrics(&(_handle->fpga_enum_metric_vector), &metric_num, _token->sysfspath, FPGA_HW_MCP);
-			if (result != FPGA_OK) {
-				FPGA_MSG("Failed to Enum Power metrics.");
+			memset_s(metrics_path, SYSFS_PATH_MAX, 0);
+
+			if (sysfs_get_fme_pwr_path(_token, metrics_path) == FPGA_OK) {
+
+				result = enum_powermgmt_metrics(&(_handle->fpga_enum_metric_vector), &metric_num, metrics_path, FPGA_HW_MCP);
+				if (result != FPGA_OK) {
+					FPGA_MSG("Failed to Enum Power metrics.");
```
Comment (tswhison) [inline libopae/plugins/xfpga/metrics/metrics_utils.c]: Seems like an error condition.  Use FPGA_ERR and return an error code?

File: libopae/plugins/xfpga/metrics/metrics_utils.c [hunk 6]
```diff
@@ -826,19 +817,32 @@ fpga_result enum_fpga_metrics(fpga_handle handle)
...
 
-			result = enum_thermalmgmt_metrics(&(_handle->fpga_enum_metric_vector), &metric_num, _token->sysfspath, FPGA_HW_MCP);
-			if (result != FPGA_OK) {
-				FPGA_MSG("Failed to Enum Thermal metrics.");
+			memset_s(metrics_path, SYSFS_PATH_MAX, 0);
+			if (sysfs_get_fme_temp_path(_token, metrics_path) == FPGA_OK) {
+
+				result = enum_thermalmgmt_metrics(&(_handle->fpga_enum_metric_vector), &metric_num, metrics_path, FPGA_HW_MCP);
+				if (result != FPGA_OK) {
+					FPGA_MSG("Failed to Enum Thermal metrics.");
```
Comment (tswhison) [inline libopae/plugins/xfpga/metrics/metrics_utils.c]: Seems like an error condition.  Use FPGA_ERR and return an error code?

File: libopae/plugins/xfpga/metrics/metrics_utils.c [hunk 7]
```diff
@@ -826,19 +817,32 @@ fpga_result enum_fpga_metrics(fpga_handle handle)
...
 
-			result = enum_perf_counter_metrics(&(_handle->fpga_enum_metric_vector), &metric_num, _token->sysfspath, FPGA_HW_MCP);
-			if (result != FPGA_OK) {
-				FPGA_MSG("Failed to Enum Perforamnce metrics.");
+			memset_s(metrics_path, SYSFS_PATH_MAX, 0);
+			if (sysfs_get_fme_perf_path(_token, metrics_path) == FPGA_OK) {
+
+				result = enum_perf_counter_metrics(&(_handle->fpga_enum_metric_vector), &metric_num, metrics_path, FPGA_HW_MCP);
+				if (result != FPGA_OK) {
+					FPGA_MSG("Failed to Enum Perforamnce metrics.");
```
Comment (tswhison) [inline libopae/plugins/xfpga/metrics/metrics_utils.c]: Seems like an error condition.  Use FPGA_ERR and return an error code?

Comment (tswhison) [inline libopae/plugins/xfpga/metrics/metrics_utils.c]: s/Perforamnce/Performance/

File: libopae/plugins/xfpga/metrics/metrics_utils.c [hunk 8]
```diff
@@ -847,20 +851,28 @@ fpga_result enum_fpga_metrics(fpga_handle handle)
...
 
-			result = enum_perf_counter_metrics(&(_handle->fpga_enum_metric_vector), &metric_num, _token->sysfspath, FPGA_HW_DCP_RC);
-			if (result != FPGA_OK) {
-				FPGA_MSG("Failed to Enum Perforamnce metrics.");
+			memset_s(metrics_path, SYSFS_PATH_MAX, 0);
+			if (sysfs_get_fme_perf_path(_token, metrics_path) == FPGA_OK) {
+
+				result = enum_perf_counter_metrics(&(_handle->fpga_enum_metric_vector), &metric_num, metrics_path, FPGA_HW_DCP_RC);
+				if (result != FPGA_OK) {
+					FPGA_MSG("Failed to Enum Perforamnce metrics.");
```
Comment (tswhison) [inline libopae/plugins/xfpga/metrics/metrics_utils.c]: Seems like an error condition.  Use FPGA_ERR and return an error code?

Comment (tswhison) [inline libopae/plugins/xfpga/metrics/metrics_utils.c]: s/Perforamnce/Performance/

File: libopae/plugins/xfpga/metrics/metrics_utils.c [hunk 9]
```diff
@@ -847,20 +851,28 @@ fpga_result enum_fpga_metrics(fpga_handle handle)
...
 
-			if (_handle->bmc_handle) {
-				result = enum_bmc_metrics_info(_handle,  &(_handle->fpga_enum_metric_vector), &metric_num, FPGA_HW_DCP_RC);
-				if (result != FPGA_OK) {
-					FPGA_MSG("Failed to enumerate BMC metrics.");
-				}
+				if (_handle->bmc_handle) {
+					result = enum_bmc_metrics_info(_handle, &(_handle->fpga_enum_metric_vector), &metric_num, FPGA_HW_DCP_RC);
+					if (result != FPGA_OK) {
+						FPGA_MSG("Failed to enumerate BMC metrics.");
```
Comment (tswhison) [inline libopae/plugins/xfpga/metrics/metrics_utils.c]: Seems like an error condition.  Use FPGA_ERR and return an error code?

File: libopae/plugins/xfpga/metrics/metrics_utils.c [hunk 10]
```diff
@@ -870,18 +882,27 @@ fpga_result enum_fpga_metrics(fpga_handle handle)
...
+
+				// Max10 Power & Thermal
+				result = enum_max10_metrics_info(_handle,
 					&(_handle->fpga_enum_metric_vector),
 					&metric_num,
 					hw_type);
-			if (result != FPGA_OK) {
-				FPGA_MSG("Failed to Enum Power and Thermal metrics.");
+				if (result != FPGA_OK) {
+					FPGA_MSG("Failed to Enum Power and Thermal metrics.");
```
Comment (tswhison) [inline libopae/plugins/xfpga/metrics/metrics_utils.c]: Seems like an error condition.  Use FPGA_ERR and return an error code?

File: libopae/plugins/xfpga/metrics/metrics_utils.c [hunk 11]
```diff
@@ -870,18 +882,27 @@ fpga_result enum_fpga_metrics(fpga_handle handle)
...
-			if (result != FPGA_OK) {
-				FPGA_MSG("Failed to Enum Performance metrics.");
+
+			memset_s(metrics_path, SYSFS_PATH_MAX, 0);
+			if (sysfs_get_fme_perf_path(_token, metrics_path) == FPGA_OK) {
+
+				// Perf Counters
+				result = enum_perf_counter_metrics(&(_handle->fpga_enum_metric_vector), &metric_num, _token->sysfspath, hw_type);
+				if (result != FPGA_OK) {
+					FPGA_MSG("Failed to Enum Performance metrics.");
```
Comment (tswhison) [inline libopae/plugins/xfpga/metrics/metrics_utils.c]: Seems like an error condition.  Use FPGA_ERR and return an error code?

File: libopae/plugins/xfpga/sysfs.c [hunk 12]
```diff
@@ -721,12 +730,22 @@ fpga_result sysfs_get_fme_pwr_path(fpga_token token, char *sysfs_pwr)
 	}
 	res = cat_token_sysfs_path(sysfs_pwr, token, SYSFS_FORMAT(sysfs_fme_pwr_glob));
 
+	// check for path is valid
+	int gres = glob(sysfs_pwr, GLOB_NOSORT, NULL, &pglob);
+	if (gres) {
+		FPGA_MSG("Failed pattern match %s: %s", sysfs_pwr, strerror(errno));
```
Comment (tswhison) [inline libopae/plugins/xfpga/sysfs.c]: use FPGA_ERR ?

File: libopae/plugins/xfpga/sysfs.c [hunk 13]
```diff
@@ -737,12 +756,22 @@ fpga_result sysfs_get_fme_temp_path(fpga_token token, char *sysfs_temp)
 
 	res = cat_token_sysfs_path(sysfs_temp, token, SYSFS_FORMAT(sysfs_fme_temp_glob));
 
+	// check for path is valid
+	int gres = glob(sysfs_temp, GLOB_NOSORT, NULL, &pglob);
+	if (gres) {
+		FPGA_MSG("Failed pattern match %s: %s", sysfs_temp, strerror(errno));
```
Comment (tswhison) [inline libopae/plugins/xfpga/sysfs.c]: Use FPGA_ERR ?

File: libopae/plugins/xfpga/sysfs.c [hunk 14]
```diff
@@ -753,6 +782,15 @@ fpga_result sysfs_get_fme_perf_path(fpga_token token, char *sysfs_perf)
 
 	res = cat_token_sysfs_path(sysfs_perf, token, SYSFS_FORMAT(sysfs_fme_perf_glob));
 
+	// check for path is valid
+	int gres = glob(sysfs_perf, GLOB_NOSORT, NULL, &pglob);
+	if (gres) {
+		FPGA_MSG("Failed pattern match %s: %s", sysfs_perf, strerror(errno));
```
Comment (tswhison) [inline libopae/plugins/xfpga/sysfs.c]: use FPGA_ERR ?

File: libopae/plugins/xfpga/metrics/metrics_utils.c [hunk 15]
```diff
@@ -293,40 +285,38 @@ fpga_result enum_powermgmt_metrics(fpga_metric_vector *vector,
...
 
-	snprintf_s_ss(sysfs_path, sizeof(sysfs_path), "%s/%s", sysfspath, PWRMGMT);
-
-	dir = opendir(sysfs_path);
-	if (NULL == dir) {
-		FPGA_MSG("can't find dir %s ", strerror(errno));
+	int gres = glob(sysfspath, GLOB_NOSORT, NULL, &pglob);
+	if (gres) {
+		FPGA_ERR("Failed pattern match %s: %s", sysfspath, strerror(errno));
+		globfree(&pglob);
```
Comment (r-rojo) [inline libopae/plugins/xfpga/metrics/metrics_utils.c]: Do we always call globfree on all errors? Are there error cases where pglob will be NULL or not freeable?
Is this the only place where we call globfree? Don't we still need it before we exit the function?

File: libopae/plugins/xfpga/metrics/metrics_utils.c (hunk 3 above)
Comment (r-rojo) [inline libopae/plugins/xfpga/metrics/metrics_utils.c]: do we clean up the vector outside of this? even on error?

File: libopae/plugins/xfpga/sysfs.c [hunk 16]
```diff
@@ -811,6 +849,47 @@ fpga_result sysfs_get_port_error_clear_path(fpga_handle handle, char *sysfs_port
...
+	}
+
+	res = cat_token_sysfs_path(sysfs_max10, token, SYSFS_FORMAT(sysfs_max10_glob));
+
+	res = opae_glob_path(sysfs_max10);
+	if (res) {
+		return res;
+	}
+
+	return res;
```
Comment (tswhison) [inline libopae/plugins/xfpga/sysfs.c]: This code can be replaced with
```c
return opae_glob_path(sysfs_max10);
```

File: libopae/plugins/xfpga/metrics/metrics_utils.c [hunk 17]
```diff
@@ -423,12 +414,10 @@ fpga_result enum_perf_counter_metrics(fpga_metric_vector *vector,
 		return FPGA_INVALID_PARAM;
 	}
 
-	snprintf_s_ss(sysfs_path, sizeof(sysfs_path), "%s/%s", sysfspath, PERF);
-
-	glob_t pglob;
-	int gres = glob(sysfs_path, GLOB_NOSORT, NULL, &pglob);
+	int gres = glob(sysfspath, GLOB_NOSORT, NULL, &pglob);
 	if ((gres) || (1 != pglob.gl_pathc)) {
 		globfree(&pglob);
```
Comment (r-rojo) [inline libopae/plugins/xfpga/metrics/metrics_utils.c]: Same thing here, should we call globfree before we exit the function?

File: libopae/plugins/xfpga/sysfs.c [hunk 18]
```diff
@@ -811,6 +849,47 @@ fpga_result sysfs_get_port_error_clear_path(fpga_handle handle, char *sysfs_port
...
+	}
+
+	res = cat_token_sysfs_path(sysfs_bmc, token, SYSFS_FORMAT(sysfs_bmc_glob));
+
+	res = opae_glob_path(sysfs_bmc);
+	if (res) {
+		return res;
+	}
+
+	return res;
```
Comment (tswhison) [inline libopae/plugins/xfpga/sysfs.c]: This code can be replaced with
```c
return opae_glob_path(sysfs_bmc);
```

File: libopae/plugins/xfpga/sysfs.c [hunk 19]
```diff
@@ -89,8 +91,11 @@ static sysfs_formats sysfs_path_table[OPAE_KERNEL_DRIVERS] = {
 	 .sysfs_compat_id = "/dfl-fme-region.*/fpga_region/region*/compat_id",
 	 .sysfs_fme_temp_glob = "hwmon/hwmon*/temp*_*",
 	 .sysfs_fme_pwr_glob = "hwmon/hwmon*/power*_*",
+	 .sysfs_fme_perf_glob = "*perf",
 	 .sysfs_port_err = "errors/errors",
-	 .sysfs_port_err_clear = "errors/errors"
+	 .sysfs_port_err_clear = "errors/errors",
+	 .sysfs_bmc_glob = "avmmi-bmc.*/bmc_info",
+	 .sysfs_max10_glob = "spi-*/spi_master/spi*/spi*.*"
```
Comment (r-rojo) [inline libopae/plugins/xfpga/sysfs.c]: does the dfl driver export sysfs nodes for spi?

File: testing/xfpga/test_sysfs_c.cpp [hunk 20]
```diff
@@ -832,19 +838,87 @@ TEST_P(sysfs_c_mock_p, fpga_sysfs_07) {
...
 *            FPGA_INVALID_PARAM <br>
 */
-TEST_P(sysfs_c_mock_p, fpga_sysfs_08) {
+TEST_P(sysfs_dfl_c_mock_p, fpga_sysfs_08) {
 	fpga_result result;
 	char sysfs_path[SYSFS_PATH_MAX] = { 0 };
 
-	result = sysfs_get_fme_perf_path(handle_, sysfs_path);
+	result = sysfs_get_fme_perf_path(tokens_[0], sysfs_path);
+	EXPECT_EQ(result, FPGA_NOT_FOUND);
```
Comment (jelonanX) [inline testing/xfpga/test_sysfs_c.cpp]: Why does this return FPGA_NOT_FOUND? Not supported in dfl 2? description doesn't match test.

File: testing/xfpga/test_sysfs_c.cpp [hunk 21]
```diff
@@ -832,19 +838,87 @@ TEST_P(sysfs_c_mock_p, fpga_sysfs_07) {
...
+	::testing::ValuesIn(test_platform::mock_platforms({ "skx-p-dfl0_patchset2" })));
+
+class sysfs_power_mock_p : public sysfs_c_mock_p { };
+/*
+* @test       sysfs
+* @brief      Tests: sysfs_get_fme_pwr_path
+ @details    When passed with valid argument return 0
+*            and fme power sysfs path <br>
+*            When passed with invalid argument return
+*            FPGA_INVALID_PARAM <br>
```
Comment (jelonanX) [inline testing/xfpga/test_sysfs_c.cpp]: description doesn't match test

File: libopae/plugins/xfpga/sysfs.c (hunk 18 above)
Comment (anandaravuri) [reply libopae/plugins/xfpga/sysfs.c]: fixed

File: libopae/plugins/xfpga/sysfs.c (hunk 14 above)
Comment (anandaravuri) [reply libopae/plugins/xfpga/sysfs.c]: fixed

File: libopae/plugins/xfpga/sysfs.c (hunk 13 above)
Comment (anandaravuri) [reply libopae/plugins/xfpga/sysfs.c]: fixed

File: libopae/plugins/xfpga/sysfs.c (hunk 12 above)
Comment (anandaravuri) [reply libopae/plugins/xfpga/sysfs.c]: fixed

File: libopae/plugins/xfpga/metrics/metrics_utils.c (hunk 5 above)
Comment (anandaravuri) [reply libopae/plugins/xfpga/metrics/metrics_utils.c]: fixed

File: libopae/plugins/xfpga/metrics/metrics_utils.c (hunk 4 above)
Comment (anandaravuri) [reply libopae/plugins/xfpga/metrics/metrics_utils.c]: fixed

File: libopae/plugins/xfpga/metrics/metrics_utils.c (hunk 3 above)
Comment (anandaravuri) [reply libopae/plugins/xfpga/metrics/metrics_utils.c]: fixed

File: libopae/plugins/xfpga/metrics/metrics_utils.c (hunk 2 above)
Comment (anandaravuri) [reply libopae/plugins/xfpga/metrics/metrics_utils.c]: fixed

File: libopae/plugins/xfpga/metrics/metrics_utils.c (hunk 1 above)
Comment (anandaravuri) [reply libopae/plugins/xfpga/metrics/metrics_utils.c]: fixed

File: libopae/plugins/xfpga/metrics/metrics_utils.c (hunk 6 above)
Comment (anandaravuri) [reply libopae/plugins/xfpga/metrics/metrics_utils.c]: fixed

File: libopae/plugins/xfpga/metrics/metrics_utils.c (hunk 7 above)
Comment (anandaravuri) [reply libopae/plugins/xfpga/metrics/metrics_utils.c]: fixed

Comment (anandaravuri) [reply libopae/plugins/xfpga/metrics/metrics_utils.c]: fixed

File: libopae/plugins/xfpga/metrics/metrics_utils.c (hunk 8 above)
Comment (anandaravuri) [reply libopae/plugins/xfpga/metrics/metrics_utils.c]: fixed

Comment (anandaravuri) [reply libopae/plugins/xfpga/metrics/metrics_utils.c]: fixed

File: libopae/plugins/xfpga/metrics/metrics_utils.c (hunk 9 above)
Comment (anandaravuri) [reply libopae/plugins/xfpga/metrics/metrics_utils.c]: fixed

File: libopae/plugins/xfpga/metrics/metrics_utils.c (hunk 11 above)
Comment (anandaravuri) [reply libopae/plugins/xfpga/metrics/metrics_utils.c]: fixed

File: testing/xfpga/test_sysfs_c.cpp (hunk 20 above)
Comment (anandaravuri) [reply testing/xfpga/test_sysfs_c.cpp]: fixed

File: testing/xfpga/test_sysfs_c.cpp (hunk 21 above)
Comment (anandaravuri) [reply testing/xfpga/test_sysfs_c.cpp]: fixed

File: libopae/plugins/xfpga/sysfs.c (hunk 19 above)
Comment (anandaravuri) [reply libopae/plugins/xfpga/sysfs.c]: DFL Driver doesn't support now, added to fix mock test cases.

File: libopae/plugins/xfpga/metrics/metrics_utils.c (hunk 17 above)
Comment (anandaravuri) [reply libopae/plugins/xfpga/metrics/metrics_utils.c]: pglob is freed in line 427

File: libopae/plugins/xfpga/metrics/metrics_utils.c (hunk 3 above)
Comment (anandaravuri) [reply libopae/plugins/xfpga/metrics/metrics_utils.c]: yes
Enum Metrics cleans vector

File: libopae/plugins/xfpga/metrics/metrics_utils.c (hunk 15 above)
Comment (anandaravuri) [reply libopae/plugins/xfpga/metrics/metrics_utils.c]: Fixed

File: libopae/plugins/xfpga/metrics/metrics_utils.c (hunk 17 above)
Comment (anandaravuri) [reply libopae/plugins/xfpga/metrics/metrics_utils.c]: fixed

File: libopae/plugins/xfpga/metrics/metrics_utils.c [hunk 22]
```diff
@@ -233,41 +231,41 @@ fpga_result enum_thermalmgmt_metrics(fpga_metric_vector *vector,
...
-		result =  get_metric_data_info(THERLGMT, dirent->d_name, mcp_metric_metadata, MCP_MDATA_SIZE, &metric_data);
+		result = get_metric_data_info(THERLGMT, (dir_name + 1), mcp_metric_metadata, MCP_MDATA_SIZE, &metric_data);
 		if (result != FPGA_OK) {
 			FPGA_MSG("Failed to get metric metadata ");
 		}
 
-		result = add_metric_vector(vector, *metric_num, THERLGMT, THERLGMT, sysfs_path, dirent->d_name, metric_sysfs, metric_data.metric_units,
-								FPGA_METRIC_DATATYPE_INT, FPGA_METRIC_TYPE_THERMAL, hw_type, 0);
+		result = add_metric_vector(vector, *metric_num, THERLGMT, THERLGMT, sysfspath, (dir_name + 1), pglob.gl_pathv[i], metric_data.metric_units,
+			FPGA_METRIC_DATATYPE_INT, FPGA_METRIC_TYPE_THERMAL, hw_type, 0);
```
Comment (r-rojo) [inline libopae/plugins/xfpga/metrics/metrics_utils.c]: Alignment is off?

File: libopae/plugins/xfpga/metrics/metrics_utils.c [hunk 23]
```diff
@@ -293,40 +289,41 @@ fpga_result enum_powermgmt_metrics(fpga_metric_vector *vector,
...
-		result = get_metric_data_info(PWRMGMT, dirent->d_name, mcp_metric_metadata, MCP_MDATA_SIZE, &metric_data);
+		result = get_metric_data_info(PWRMGMT, (dir_name + 1), mcp_metric_metadata, MCP_MDATA_SIZE, &metric_data);
 		if (result != FPGA_OK) {
 			FPGA_MSG("Failed to get metric metadata ");
 		}
 
-		result = add_metric_vector(vector, *metric_num, PWRMGMT, PWRMGMT, sysfs_path, dirent->d_name, metric_sysfs, metric_data.metric_units,
-									FPGA_METRIC_DATATYPE_INT, FPGA_METRIC_TYPE_POWER, hw_type, 0);
+		result = add_metric_vector(vector, *metric_num, PWRMGMT, PWRMGMT, sysfspath, (dir_name + 1), pglob.gl_pathv[i], metric_data.metric_units,
+			FPGA_METRIC_DATATYPE_INT, FPGA_METRIC_TYPE_POWER, hw_type, 0);
```
Comment (r-rojo) [inline libopae/plugins/xfpga/metrics/metrics_utils.c]: Alignment off?

File: libopae/plugins/xfpga/sysfs.c (hunk 19 above)
Comment (r-rojo) [reply libopae/plugins/xfpga/sysfs.c]: Should they be null then?

File: libopae/plugins/xfpga/sysfs.c [hunk 24]
```diff
@@ -720,13 +729,26 @@ fpga_result sysfs_get_fme_pwr_path(fpga_token token, char *sysfs_pwr)
 		return FPGA_INVALID_PARAM;
 	}
 	res = cat_token_sysfs_path(sysfs_pwr, token, SYSFS_FORMAT(sysfs_fme_pwr_glob));
+	if (res != FPGA_OK) {
+		return res;
+	}
+
+	// check for path is valid
+	int gres = glob(sysfs_pwr, GLOB_NOSORT, NULL, &pglob);
+	if (gres) {
```
Comment (r-rojo) [inline libopae/plugins/xfpga/sysfs.c]: We should probably if glob returns GLOB_NOSPACE and translate that into an FPGA_NO_MEMORY result

File: libopae/plugins/xfpga/sysfs.c [hunk 25]
```diff
@@ -720,13 +729,26 @@ fpga_result sysfs_get_fme_pwr_path(fpga_token token, char *sysfs_pwr)
...
+		FPGA_ERR("Failed pattern match %s: %s", sysfs_pwr, strerror(errno));
+		globfree(&pglob);
```
Comment (r-rojo) [inline libopae/plugins/xfpga/sysfs.c]: globfree may fail the underlying buffer is never allocated.

File: libopae/plugins/xfpga/sysfs.c [hunk 26]
```diff
@@ -736,13 +758,26 @@ fpga_result sysfs_get_fme_temp_path(fpga_token token, char *sysfs_temp)
 	}
 
 	res = cat_token_sysfs_path(sysfs_temp, token, SYSFS_FORMAT(sysfs_fme_temp_glob));
+	if (res != FPGA_OK) {
+		return res;
+	}
+
+	// check for path is valid
+	int gres = glob(sysfs_temp, GLOB_NOSORT, NULL, &pglob);
```
Comment (r-rojo) [inline libopae/plugins/xfpga/sysfs.c]: We should probably if glob returns GLOB_NOSPACE and translate that into an FPGA_NO_MEMORY result
Maybe translate GLOB_NOMATCH to FPGA_NOT_FOUND and GLOB_ABORTED into FPGA_EXCEPTION?

Comment (anandaravuri) [reply libopae/plugins/xfpga/sysfs.c]: Fiexd

File: libopae/plugins/xfpga/sysfs.c (hunk 25 above)
Comment (anandaravuri) [reply libopae/plugins/xfpga/sysfs.c]: fiexd

File: libopae/plugins/xfpga/sysfs.c (hunk 24 above)
Comment (anandaravuri) [reply libopae/plugins/xfpga/sysfs.c]: fixed

File: libopae/plugins/xfpga/metrics/metrics_utils.c (hunk 23 above)
Comment (anandaravuri) [reply libopae/plugins/xfpga/metrics/metrics_utils.c]: fixed

File: libopae/plugins/xfpga/metrics/metrics_utils.c (hunk 22 above)
Comment (anandaravuri) [reply libopae/plugins/xfpga/metrics/metrics_utils.c]: fixed

File: libopae/plugins/xfpga/sysfs.c [hunk 27]
```diff
@@ -795,10 +811,26 @@ fpga_result sysfs_get_fme_perf_path(fpga_token token, char *sysfs_perf)
...
+			res = FPGA_EXCEPTION;
+		}
+
+		if (pglob.gl_pathv) {
+			globfree(&pglob);
+		}
+		return res;
+	}
+
+	if (pglob.gl_pathv) {
```
Comment (jelonanX) [inline libopae/plugins/xfpga/sysfs.c]: seems like free glob code is duplicated here?

File: libopae/plugins/xfpga/sysfs.c [hunk 28]
```diff
@@ -766,10 +766,26 @@ fpga_result sysfs_get_fme_temp_path(fpga_token token, char *sysfs_temp)
...
```
Comment (jelonanX) [inline libopae/plugins/xfpga/sysfs.c]: seems like free glob code is duplicated here?

File: libopae/plugins/xfpga/metrics/metrics_utils.c [hunk 29]
```diff
@@ -234,6 +234,7 @@ fpga_result enum_thermalmgmt_metrics(fpga_metric_vector *vector,
 	int gres = glob(sysfspath, GLOB_NOSORT, NULL, &pglob);
 	if (gres) {
 		FPGA_ERR("Failed pattern match %s: %s", sysfspath, strerror(errno));
+		//TBD refactor to common function
```
Comment (r-rojo) [inline libopae/plugins/xfpga/metrics/metrics_utils.c]: You mean // TODO: ...

**Pull request merged**